File: packages/ash-php/tests/SecurityComprehensiveTest.php
<?php

declare(strict_types=1);

namespace Ash\Tests;

use Ash\Ash;
use Ash\Core\Compare;
use Ash\Core\Proof;
use PHPUnit\Framework\Attributes\Test;
use PHPUnit\Framework\TestCase;

/**
 * Comprehensive security-focused tests.
 */
final class SecurityComprehensiveTest extends TestCase
{
    // ========== TIMING ATTACK RESISTANCE ==========

    #[Test]
    public function timingSafeCompareResistsTimingAttack(): void
    {
        $secret = hash('sha256', 'actual_secret');

        // Different length should return false
        $this->assertFalse(Compare::timingSafe($secret, 'short'));

        // First char wrong
        $this->assertFalse(Compare::timingSafe($secret, 'x' . substr($secret, 1)));

        // Last char wrong
        $this->assertFalse(Compare::timingSafe($secret, substr($secret, 0, -1) . 'x'));

        // Middle char wrong
        $this->assertFalse(Compare::timingSafe($secret, substr($secret, 0, 32) . 'x' . substr($secret, 33)));

        // Correct comparison
        $this->assertTrue(Compare::timingSafe($secret, $secret));
    }

    #[Test]
    public function timingSafeCompareDifferentPositions(): void
    {
        $original = str_repeat('a', 100);

        for ($pos = 0; $pos < 100; $pos++) {
            $modified = substr($original, 0, $pos) . 'b' . substr($original, $pos + 1);
            $this->assertFalse(Compare::timingSafe($original, $modified));
        }
    }

    // ========== PROOF TAMPERING DETECTION ==========

    #[Test]
    public function detectsSingleBitChange(): void
    {
        $nonce = str_repeat('a', 64);
        $contextId = 'ctx_test';
        $binding = 'POST|/api|';
        $timestamp = '1234567890';
        $bodyHash = Proof::hashBody('data');

        $clientSecret = Proof::deriveClientSecret($nonce, $contextId, $binding);
        $proof = Proof::ashBuildProofHmac($clientSecret, $timestamp, $binding, $bodyHash);

        // Change single character
        $chars = str_split($proof);
        for ($i = 0; $i < strlen($proof); $i++) {
            $tampered = $chars;
            $tampered[$i] = $chars[$i] === 'a' ? 'b' : 'a';
            $tamperedProof = implode('', $tampered);

            $this->assertFalse(
                Proof::ashVerifyProof($nonce, $contextId, $binding, $timestamp, $bodyHash, $tamperedProof),
                "Should detect tampering at position $i"
            );
        }
    }

    #[Test]
    public function detectsTimestampTampering(): void
    {
        $nonce = str_repeat('b', 64);
        $contextId = 'ctx_timestamp';
        $binding = 'POST|/api|';
        $bodyHash = Proof::hashBody('data');

        $originalTimestamp = '1234567890';
        $clientSecret = Proof::deriveClientSecret($nonce, $contextId, $binding);
        $proof = Proof::ashBuildProofHmac($clientSecret, $originalTimestamp, $binding, $bodyHash);

        // Try different timestamps
        $tamperedTimestamps = [
            '1234567891', // Off by one
            '0000000000', // All zeros
            '9999999999', // All nines
            '', // Empty
            '-1', // Negative
        ];

        foreach ($tamperedTimestamps as $tampered) {
            $this->assertFalse(
                Proof::ashVerifyProof($nonce, $contextId, $binding, $tampered, $bodyHash, $proof),
                "Should detect timestamp tampering: $tampered"
            );
        }
    }

    #[Test]
    public function detectsBindingTampering(): void
    {
        $nonce = str_repeat('c', 64);
        $contextId = 'ctx_binding';
        $timestamp = '1234567890';
        $bodyHash = Proof::hashBody('data');

        $originalBinding = 'POST|/api/users|';
        $clientSecret = Proof::deriveClientSecret($nonce, $contextId, $originalBinding);
        $proof = Proof::ashBuildProofHmac($clientSecret, $timestamp, $originalBinding, $bodyHash);

        // Try different bindings - verifying with different binding should fail
        $tamperedBindings = [
            'GET|/api/users|', // Wrong method
            'POST|/api/admin|', // Wrong path
            'POST|/api/users|extra', // Extra data
            '', // Empty
        ];

        foreach ($tamperedBindings as $tampered) {
            $this->assertFalse(
                Proof::ashVerifyProof($nonce, $contextId, $tampered, $timestamp, $bodyHash, $proof),
                "Should detect binding tampering: $tampered"
            );
        }
    }

    #[Test]
    public function detectsBodyTampering(): void
    {
        $nonce = str_repeat('d', 64);
        $contextId = 'ctx_body';
        $timestamp = '1234567890';
        $binding = 'POST|/api|';

        $originalBody = '{"name":"John","action":"transfer"}';
        $originalHash = Proof::hashBody($originalBody);
        $clientSecret = Proof::deriveClientSecret($nonce, $contextId, $binding);
        $proof = Proof::ashBuildProofHmac($clientSecret, $timestamp, $binding, $originalHash);

        // Try tampered bodies
        $tamperedBodies = [
            '{"name":"Jane","action":"transfer"}', // Changed name
            '{"name":"John","action":"delete"}', // Changed action
            '{"name":"John","action":"transfer","extra":true}', // Added field
            '', // Empty
        ];

        foreach ($tamperedBodies as $tampered) {
            $tamperedHash = Proof::hashBody($tampered);
            $this->assertFalse(
                Proof::ashVerifyProof($nonce, $contextId, $binding, $timestamp, $tamperedHash, $proof),
                "Should detect body tampering"
            );
        }
    }

    // ========== REPLAY ATTACK RESISTANCE ==========

    #[Test]
    public function sameInputsProduceSameProof(): void
    {
        // This is a property test - same inputs = same output
        // Replay detection must be handled at the context level
        $secret = 'secret';
        $timestamp = '1234567890';
        $binding = 'POST|/api|';
        $bodyHash = 'abc123';

        $proof1 = Proof::ashBuildProofHmac($secret, $timestamp, $binding, $bodyHash);
        $proof2 = Proof::ashBuildProofHmac($secret, $timestamp, $binding, $bodyHash);

        $this->assertSame($proof1, $proof2);
    }

    #[Test]
    public function differentContextsProduceDifferentProofs(): void
    {
        // Nonce/context should be included to prevent replay
        $timestamp = '1234567890';
        $binding = 'POST|/api|';
        $bodyHash = 'abc123';

        $secret1 = Proof::deriveClientSecret('11111111111111111111111111111111', 'ctx1', $binding);
        $secret2 = Proof::deriveClientSecret('22222222222222222222222222222222', 'ctx2', $binding);

        $proof1 = Proof::ashBuildProofHmac($secret1, $timestamp, $binding, $bodyHash);
        $proof2 = Proof::ashBuildProofHmac($secret2, $timestamp, $binding, $bodyHash);

        $this->assertNotSame($proof1, $proof2);
    }

    // ========== SECRET DERIVATION SECURITY ==========

    #[Test]
    public function clientSecretIsNotReversible(): void
    {
        $secret = Proof::deriveClientSecret('abcd1234abcd1234abcd1234abcd1234', 'ctx456', 'POST|/api|');

        // Secret should not contain input values
        $this->assertStringNotContainsString('abcd1234abcd1234abcd1234abcd1234', $secret);
        $this->assertStringNotContainsString('ctx456', $secret);
        $this->assertStringNotContainsString('nonce', $secret);
        $this->assertStringNotContainsString('ctx', $secret);
    }

    #[Test]
    public function clientSecretHasHighEntropy(): void
    {
        $secrets = [];

        for ($i = 0; $i < 100; $i++) {
            $secrets[] = Proof::deriveClientSecret(sprintf('%032x', $i) . sprintf('%032x', $i + 1000), sprintf('%032x', $i) . sprintf('%032x', $i + 2000), 'POST|/api|');
        }

        // All should be unique
        $unique = array_unique($secrets);
        $this->assertCount(100, $unique);

        // Check character distribution
        $allChars = implode('', $secrets);
        $hexChars = '0123456789abcdef';

        foreach (str_split($hexChars) as $char) {
            $count = substr_count($allChars, $char);
            // Each hex char should appear roughly equally (6400 chars / 16 = 400)
            $this->assertGreaterThan(200, $count, "Character $char appears too rarely");
        }
    }

    // ========== HASH SECURITY ==========

    #[Test]
    public function hashResistsCollisions(): void
    {
        $hashes = [];

        for ($i = 0; $i < 1000; $i++) {
            $hashes[] = Proof::hashBody("input_$i");
        }

        $unique = array_unique($hashes);
        $this->assertCount(1000, $unique);
    }

    #[Test]
    public function hashAvalancheEffect(): void
    {
        $hash1 = Proof::hashBody('test');
        $hash2 = Proof::hashBody('test1'); // One char added

        // Count different characters
        $diffCount = 0;
        for ($i = 0; $i < 64; $i++) {
            if ($hash1[$i] !== $hash2[$i]) {
                $diffCount++;
            }
        }

        // Should have significant difference (avalanche effect)
        $this->assertGreaterThan(20, $diffCount);
    }

    #[Test]
    public function hashPreimageResistance(): void
    {
        $hash = Proof::hashBody('secret_data');

        // Hash should not reveal input
        $this->assertStringNotContainsString('secret', $hash);
        $this->assertStringNotContainsString('data', $hash);
    }

    // ========== INPUT VALIDATION SECURITY ==========

    #[Test]
    public function handlesNullByteAttack(): void
    {
        // Null byte shouldn't truncate strings
        $hash1 = Proof::hashBody("test\x00data");
        $hash2 = Proof::hashBody("test");

        $this->assertNotSame($hash1, $hash2);
    }

    #[Test]
    public function handlesOverlongUtf8(): void
    {
        // Test that overlong UTF-8 sequences are handled safely
        $normal = 'test';
        $hash1 = Proof::hashBody($normal);

        // Should not crash or behave unexpectedly
        $this->assertSame(64, strlen($hash1));
    }

    // ========== SCOPED PROOF SECURITY ==========

    #[Test]
    public function scopedProofProtectsSelectedFields(): void
    {
        $nonce = str_repeat('e', 64);
        $contextId = 'ctx_scoped';
        $binding = 'POST|/api|';
        $timestamp = '1234567890';

        $payload = ['name' => 'John', 'email' => 'john@example.com'];
        $scope = ['name'];

        $clientSecret = Proof::deriveClientSecret($nonce, $contextId, $binding);
        $result = Proof::ashBuildProofScoped($clientSecret, $timestamp, $binding, $payload, $scope);

        // Changing scoped field should break verification
        $tamperedPayload = ['name' => 'Jane', 'email' => 'john@example.com'];
        $verified = Proof::ashVerifyProofScoped(
            $nonce, $contextId, $binding, $timestamp, $tamperedPayload, $scope,
            $result['scopeHash'], $result['proof']
        );
        $this->assertFalse($verified);
    }

    #[Test]
    public function scopedProofAllowsUnscopedChanges(): void
    {
        $nonce = str_repeat('f', 64);
        $contextId = 'ctx_unscoped';
        $binding = 'POST|/api|';
        $timestamp = '1234567890';

        $payload = ['name' => 'John', 'email' => 'john@example.com'];
        $scope = ['name'];

        $clientSecret = Proof::deriveClientSecret($nonce, $contextId, $binding);
        $result = Proof::ashBuildProofScoped($clientSecret, $timestamp, $binding, $payload, $scope);

        // Changing unscoped field should still verify
        $modifiedPayload = ['name' => 'John', 'email' => 'different@example.com'];
        $verified = Proof::ashVerifyProofScoped(
            $nonce, $contextId, $binding, $timestamp, $modifiedPayload, $scope,
            $result['scopeHash'], $result['proof']
        );
        $this->assertTrue($verified);
    }

    // ========== CHAIN INTEGRITY ==========

    #[Test]
    public function chainHashPreventsReordering(): void
    {
        $nonce = str_repeat('a', 64);
        $contextId = 'ctx_chain';
        $binding1 = 'POST|/api/step1|';
        $binding2 = 'POST|/api/step2|';

        $clientSecret = Proof::deriveClientSecret($nonce, $contextId, $binding1);

        // Create chain of requests
        $result1 = Proof::ashBuildProofUnified(
            $clientSecret, '1000000001', $binding1,
            ['step' => 1], ['step'], null
        );

        $result2 = Proof::ashBuildProofUnified(
            $clientSecret, '1000000002', $binding2,
            ['step' => 2], ['step'], $result1['proof']
        );

        // Chain hash should match previous proof hash
        $expectedChainHash = Proof::hashProof($result1['proof']);
        $this->assertSame($expectedChainHash, $result2['chainHash']);
    }

    // ========== NONCE GENERATION SECURITY ==========

    #[Test]
    public function nonceHasSufficientEntropy(): void
    {
        $nonces = [];

        for ($i = 0; $i < 1000; $i++) {
            $nonces[] = Ash::generateNonce();
        }

        // All should be unique
        $unique = array_unique($nonces);
        $this->assertCount(1000, $unique);
    }

    #[Test]
    public function nonceHasMinimumLength(): void
    {
        for ($i = 0; $i < 100; $i++) {
            $nonce = Ash::generateNonce();
            $this->assertGreaterThanOrEqual(32, strlen($nonce)); // 16 bytes = 32 hex
        }
    }

    // ========== CONTEXT ID GENERATION SECURITY ==========

    #[Test]
    public function contextIdIsUnique(): void
    {
        $ids = [];

        for ($i = 0; $i < 1000; $i++) {
            $ids[] = Ash::generateContextId();
        }

        $unique = array_unique($ids);
        $this->assertCount(1000, $unique);
    }

    // ========== BASE64URL SECURITY ==========

    #[Test]
    public function base64UrlIsUrlSafe(): void
    {
        // Test data that would produce unsafe chars in standard base64
        $testData = [
            "\xfb\xff\xfe", // Would produce + and / in standard
            random_bytes(100),
            str_repeat("\x00\xff", 50),
        ];

        foreach ($testData as $data) {
            $encoded = Proof::base64UrlEncode($data);

            $this->assertStringNotContainsString('+', $encoded);
            $this->assertStringNotContainsString('/', $encoded);
            $this->assertStringNotContainsString('=', $encoded);

            // Should round-trip correctly
            $decoded = Proof::base64UrlDecode($encoded);
            $this->assertSame($data, $decoded);
        }
    }
}

File: packages/ash-go/security_comprehensive_test.go
package ash

import (
	"fmt"
	"strings"
	"testing"
)

// Security-focused tests

// TestTimingAttackResistanceFirstCharDiff tests timing safe compare with first char different
func TestTimingAttackResistanceFirstCharDiff(t *testing.T) {
	secret := AshHashBody("actual_secret")
	modified := "x" + secret[1:]
	if AshTimingSafeCompare(secret, modified) {
		t.Error("Should detect first char difference")
	}
}

// TestTimingAttackResistanceLastCharDiff tests timing safe compare with last char different
func TestTimingAttackResistanceLastCharDiff(t *testing.T) {
	secret := AshHashBody("actual_secret")
	modified := secret[:len(secret)-1] + "x"
	if AshTimingSafeCompare(secret, modified) {
		t.Error("Should detect last char difference")
	}
}

// TestTimingAttackResistanceMiddleCharDiff tests timing safe compare with middle char different
func TestTimingAttackResistanceMiddleCharDiff(t *testing.T) {
	secret := AshHashBody("actual_secret")
	modified := secret[:32] + "x" + secret[33:]
	if AshTimingSafeCompare(secret, modified) {
		t.Error("Should detect middle char difference")
	}
}

// TestDetectsSingleBitChangeInProof tests detection of single bit change
func TestDetectsSingleBitChangeInProofSecurity(t *testing.T) {
	secret := "test-secret"
	timestamp := "1234567890"
	binding := "POST|/api|"
	bodyHash := AshHashBody("data")

	proof := AshBuildProofHMAC(secret, timestamp, binding, bodyHash)

	chars := []byte(proof)
	for i := 0; i < len(proof); i++ {
		tampered := make([]byte, len(chars))
		copy(tampered, chars)
		if tampered[i] == 'a' {
			tampered[i] = 'b'
		} else {
			tampered[i] = 'a'
		}
		tamperedProof := string(tampered)

		valid, err := AshVerifyProof("abcd1234abcd1234abcd1234abcd1234", "ctx123456789012345678901234567890", binding, timestamp, bodyHash, tamperedProof)
		if err != nil {
			t.Fatal(err)
		}
		if valid {
			t.Errorf("Should detect tampering at position %d", i)
		}
	}
}

// TestDetectsTimestampTamperingSecurity tests detection of timestamp tampering
func TestDetectsTimestampTamperingSecurity(t *testing.T) {
	binding := "POST|/api|"
	bodyHash := AshHashBody("data")

	originalTimestamp := "1234567890"
	nonce := "abcd1234abcd1234abcd1234abcd1234"
	contextID := "ctx123"
	clientSecret, err := AshDeriveClientSecret(nonce, contextID, binding)
	if err != nil {
		t.Fatal(err)
	}
	proof := AshBuildProofHMAC(clientSecret, originalTimestamp, binding, bodyHash)

	tamperedTimestamps := []string{
		"1234567891",
		"0000000000",
		"9999999999",
	}

	for _, tampered := range tamperedTimestamps {
		valid, err := AshVerifyProof(nonce, contextID, binding, tampered, bodyHash, proof)
		if err != nil {
			t.Fatal(err)
		}
		if valid {
			t.Errorf("Should detect timestamp tampering: %s", tampered)
		}
	}
}

// TestDetectsBindingTamperingSecurity tests detection of binding tampering
func TestDetectsBindingTamperingSecurity(t *testing.T) {
	timestamp := "1234567890"
	bodyHash := AshHashBody("data")

	originalBinding := "POST|/api/users|"
	nonce := "abcd1234abcd1234abcd1234abcd1234"
	contextID := "ctx123"
	clientSecret, err := AshDeriveClientSecret(nonce, contextID, originalBinding)
	if err != nil {
		t.Fatal(err)
	}
	proof := AshBuildProofHMAC(clientSecret, timestamp, originalBinding, bodyHash)

	tamperedBindings := []string{
		"GET|/api/users|",
		"POST|/api/admin|",
		"POST|/api/users|extra",
	}

	for _, tampered := range tamperedBindings {
		valid, err := AshVerifyProof(nonce, contextID, tampered, timestamp, bodyHash, proof)
		if err != nil {
			t.Fatal(err)
		}
		if valid {
			t.Errorf("Should detect binding tampering: %s", tampered)
		}
	}
}

// TestDetectsBodyTamperingSecurity tests detection of body tampering
func TestDetectsBodyTamperingSecurity(t *testing.T) {
	timestamp := "1234567890"
	binding := "POST|/api|"

	originalBody := `{"name":"John","action":"transfer"}`
	originalHash := AshHashBody(originalBody)
	nonce := "abcd1234abcd1234abcd1234abcd1234"
	contextID := "ctx123"
	clientSecret, err := AshDeriveClientSecret(nonce, contextID, binding)
	if err != nil {
		t.Fatal(err)
	}
	proof := AshBuildProofHMAC(clientSecret, timestamp, binding, originalHash)

	tamperedBodies := []string{
		`{"name":"Jane","action":"transfer"}`,
		`{"name":"John","action":"delete"}`,
		`{"name":"John","action":"transfer","extra":true}`,
	}

	for _, tampered := range tamperedBodies {
		tamperedHash := AshHashBody(tampered)
		valid, err := AshVerifyProof(nonce, contextID, binding, timestamp, tamperedHash, proof)
		if err != nil {
			t.Fatal(err)
		}
		if valid {
			t.Error("Should detect body tampering")
		}
	}
}

// TestReplayAttackSameInputsSameProofSecurity tests that same inputs produce same proof
func TestReplayAttackSameInputsSameProofSecurity(t *testing.T) {
	secret := "secret"
	timestamp := "1234567890"
	binding := "POST|/api|"
	bodyHash := "abc123"

	proof1 := AshBuildProofHMAC(secret, timestamp, binding, bodyHash)
	proof2 := AshBuildProofHMAC(secret, timestamp, binding, bodyHash)

	if proof1 != proof2 {
		t.Error("Same inputs should produce same proof")
	}
}

// TestDifferentContextsDifferentProofsSecurity tests different contexts produce different proofs
func TestDifferentContextsDifferentProofsSecurity(t *testing.T) {
	timestamp := "1234567890"
	binding := "POST|/api|"
	bodyHash := "abc123"

	secret1, err := AshDeriveClientSecret("abcd1234abcd1234abcd1234abcd1234", "ctx123456789012345678901234567890", binding)
	if err != nil {
		t.Fatal(err)
	}
	secret2, err := AshDeriveClientSecret("bcde2345bcde2345bcde2345bcde2345", "ctx234567890123456789012345678901", binding)
	if err != nil {
		t.Fatal(err)
	}

	proof1 := AshBuildProofHMAC(secret1, timestamp, binding, bodyHash)
	proof2 := AshBuildProofHMAC(secret2, timestamp, binding, bodyHash)

	if proof1 == proof2 {
		t.Error("Different contexts should produce different proofs")
	}
}

// TestClientSecretIsNotReversibleSecurity tests client secret is not reversible
func TestClientSecretIsNotReversibleSecurity(t *testing.T) {
	secret, err := AshDeriveClientSecret("abcd1234abcd1234abcd1234abcd1234", "ctx456789012345678901234567890123", "binding")
	if err != nil {
		t.Fatal(err)
	}

	if strings.Contains(secret, "nonce123") {
		t.Error("Secret should not contain nonce")
	}
	if strings.Contains(secret, "ctx456") {
		t.Error("Secret should not contain context")
	}
}

// TestClientSecretHasHighEntropySecurity tests client secret has high entropy
func TestClientSecretHasHighEntropySecurity(t *testing.T) {
	secrets := make(map[string]bool)

	for i := 0; i < 100; i++ {
		secret, err := AshDeriveClientSecret(fmt.Sprintf("abcd1234abcd1234abcd1234abcd12%02d", i), fmt.Sprintf("ctx123456789012345678901234567%03d", i), "binding")
		if err != nil {
			t.Fatal(err)
		}
		if secrets[secret] {
			t.Error("Duplicate secret found")
		}
		secrets[secret] = true
	}

	if len(secrets) != 100 {
		t.Error("All secrets should be unique")
	}
}

// TestHashResistsCollisionsSecurity tests hash resists collisions
func TestHashResistsCollisionsSecurity(t *testing.T) {
	hashes := make(map[string]bool)

	for i := 0; i < 1000; i++ {
		hash := AshHashBody(fmt.Sprintf("input_%d", i))
		if hashes[hash] {
			t.Error("Hash collision found")
		}
		hashes[hash] = true
	}
}

// TestHashAvalancheEffectSecurity tests hash avalanche effect
func TestHashAvalancheEffectSecurity(t *testing.T) {
	hash1 := AshHashBody("test")
	hash2 := AshHashBody("test1")

	diffCount := 0
	for i := 0; i < 64; i++ {
		if hash1[i] != hash2[i] {
			diffCount++
		}
	}

	if diffCount < 20 {
		t.Errorf("Avalanche effect not sufficient: only %d chars different", diffCount)
	}
}

// TestHashPreimageResistanceSecurity tests hash preimage resistance
func TestHashPreimageResistanceSecurity(t *testing.T) {
	hash := AshHashBody("secret_data")

	if strings.Contains(hash, "secret") {
		t.Error("Hash should not reveal input")
	}
	if strings.Contains(hash, "data") {
		t.Error("Hash should not reveal input")
	}
}

// TestNullByteAttackSecurity tests null byte attack
func TestNullByteAttackSecurity(t *testing.T) {
	hash1 := AshHashBody("test\x00data")
	hash2 := AshHashBody("test")

	if hash1 == hash2 {
		t.Error("Null byte should not truncate")
	}
}

// TestNonceHasSufficientEntropySecurity tests nonce has sufficient entropy
func TestNonceHasSufficientEntropySecurity(t *testing.T) {
	nonces := make(map[string]bool)

	for i := 0; i < 1000; i++ {
		nonce, err := AshGenerateNonce(16)
		if err != nil {
			t.Errorf("Error generating nonce: %v", err)
		}
		if nonces[nonce] {
			t.Error("Duplicate nonce found")
		}
		nonces[nonce] = true
	}
}

// TestNonceHasMinimumLengthSecurity tests nonce has minimum length
func TestNonceHasMinimumLengthSecurity(t *testing.T) {
	for i := 0; i < 100; i++ {
		nonce, err := AshGenerateNonce(16)
		if err != nil {
			t.Errorf("Error generating nonce: %v", err)
		}
		if len(nonce) < 32 {
			t.Errorf("Nonce too short: %d", len(nonce))
		}
	}
}

// TestContextIdIsUniqueSecurity tests context ID is unique
func TestContextIdIsUniqueSecurity(t *testing.T) {
	ids := make(map[string]bool)

	for i := 0; i < 1000; i++ {
		id, err := AshGenerateContextID()
		if err != nil {
			t.Errorf("Error generating context ID: %v", err)
		}
		if ids[id] {
			t.Error("Duplicate context ID found")
		}
		ids[id] = true
	}
}

// TestBase64UrlIsUrlSafeSecurity tests base64url is url safe
func TestBase64UrlIsUrlSafeSecurity(t *testing.T) {
	testData := [][]byte{
		{0xfb, 0xff, 0xfe},
		{0x00, 0xff, 0x7f, 0x80},
	}

	for _, data := range testData {
		encoded := Base64URLEncode(data)

		if strings.Contains(encoded, "+") {
			t.Error("Base64URL should not contain +")
		}
		if strings.Contains(encoded, "/") {
			t.Error("Base64URL should not contain /")
		}
		if strings.Contains(encoded, "=") {
			t.Error("Base64URL should not contain =")
		}

		decoded, err := Base64URLDecode(encoded)
		if err != nil {
			t.Errorf("Decode error: %v", err)
		}
		if string(decoded) != string(data) {
			t.Error("Round trip failed")
		}
	}
}

// TestProofHasGoodDistributionSecurity tests proof has good distribution
func TestProofHasGoodDistributionSecurity(t *testing.T) {
	charCounts := make(map[rune]int)
	for _, c := range "0123456789abcdef" {
		charCounts[c] = 0
	}

	for i := 0; i < 100; i++ {
		proof := AshBuildProofHMAC(
			fmt.Sprintf("secret_%d", i),
			"1000000000",
			fmt.Sprintf("POST|/api/%d|", i),
			fmt.Sprintf("body_%d", i),
		)

		for _, c := range proof {
			charCounts[c]++
		}
	}

	for c, count := range charCounts {
		if count < 200 {
			t.Errorf("Character %c appears too infrequently: %d", c, count)
		}
		if count > 600 {
			t.Errorf("Character %c appears too frequently: %d", c, count)
		}
	}
}

// TestHmacProducesFixedLengthOutputSecurity tests HMAC produces fixed length output
func TestHmacProducesFixedLengthOutputSecurity(t *testing.T) {
	lengths := []int{1, 10, 100, 1000, 10000}

	for _, l := range lengths {
		proof := AshBuildProofHMAC(
			strings.Repeat("s", l),
			"1234567890",
			"POST|/api|",
			strings.Repeat("a", 64),
		)
		if len(proof) != 64 {
			t.Errorf("Expected 64 chars for secret length %d, got %d", l, len(proof))
		}
	}
}

// TestSmallInputChangeProducesDifferentOutputSecurity tests small input change produces different output
func TestSmallInputChangeProducesDifferentOutputSecurity(t *testing.T) {
	proof1 := AshBuildProofHMAC("secret", "1234567890", "POST|/api|", "abc123")
	proof2 := AshBuildProofHMAC("secret", "1234567891", "POST|/api|", "abc123")

	if proof1 == proof2 {
		t.Error("Different inputs should produce different outputs")
	}

	diffCount := 0
	for i := 0; i < len(proof1); i++ {
		if proof1[i] != proof2[i] {
			diffCount++
		}
	}

	if diffCount < 20 {
		t.Errorf("Avalanche effect not sufficient: %d chars different", diffCount)
	}
}
